tests/test_install.py

```diff
@@ -122,6 +127,48 @@ def test_calls_func(self):
         func.assert_called_with(some_path)
 
 
+@pytest.mark.parametrize("args", [
+    (),
+    ("one", ),
+    ("one", "two", "three", ),
+])
+def test_ensure_write_takes_two_arguments(args):
+    with pytest.raises(TypeError):
+        install.ensure_write(*args)
+
+
+class ensure_write_TestCase(unittest.TestCase):
+    def setUp(self):
+        self.prefix = generate_random_path()
+        self.dist = "foobar-%s-0" % random.randint(100, 200)
+        self.meta = {
+            "name": self.dist,
+            "files": ["a", "b", "c"],
+        }
+
+    @contextmanager
+    def generate_mock_check(self, return_value=True):
+        with patch.object(install, 'can_open_all_files_in_prefix') as mocked:
+            mocked.return_value = return_value
+            with patch.object(install, "load_meta") as load_meta:
+                load_meta.return_value = self.meta
+                yield mocked
+
+    def test_dispatches_to_can_open_all_files(self):
+        with self.generate_mock_check() as can:
+            install.ensure_write(self.prefix, self.dist)
+        can.assert_called_with(self.prefix, self.meta['files'])
+
+    def test_returns_none_on_true(self):
+        with self.generate_mock_check():
+            actual = install.ensure_write(self.prefix, self.meta)
```

Comment:
Sorry -- been out of the office for the better part of the last week and just realized it was only on my calendar. :-(

----

This test is actually misleading as it should be `self.prefix, 'some_pkg-1.0.0-0`, not the meta.  Sorry for writing it that way.  `load_meta` requires a dist string, not the meta data since it actually loads the meta data given a string when used, but in this case it's mocked out inside `generate_mock_check` and it always returns `self.meta` without checking what was passed in.

Looking through the commit log associated with `generate_mock_check`, I see 2099eb7721263b168b56f60b4c67805d4fa2363b that changed the behavior.  Additional detail not captured in that commit log (that should have been) are that the original assumption was that by this point I had a prefix string and a dictionary representing a package.  That was wrong, so the test was adjusted, but not the parameters passed in to `ensure_write`.  Since everything was mocked, it doesn't make much difference other than to confuse folks who come along behind me.

---

Was there anything else in this code that was confusing.  Reading through it I can see the following steps:

* Pass out to a context manager which
  * Mocks `can_open_all_files_in_prefix` and sets it to return `True`
  * Mocks `load_meta` and sets it to always return `self.meta` regardless of what it is provided
  * `yield`s back
* Calls `ensure_write` which only calls those two functions which now have the mocked behavior described above

I purposely try to keep test cases very small and compact and the implementation of the function under test is likewise small (it calls two functions and throws an exception if the second returns `False`) to make it easier to understand.  Passing in invalid parameters definitely didn't help here, but I would love some ideas on how I could have made this more approachable.